perfrunbook/system-load-and-compute-headroom.md

```diff
@@ -0,0 +1,157 @@
+# System Load and Compute Headroom
+
+The typical pricing model for cloud computing services centers on the amount of hardware resources set aside
+for the customer. With this, ideally such resources should be utilized fully, to operate at peak economy.
+In the ideal world the performance of a system should be linear up to 100% load.
+This can be in the context of pure compute or peak load and fail-over resilience.
+However, depending on workloads, systems that employ simultaneous multithreading (SMT), 
+aka 'Hyperthreading', incur non-linear performance degradation when the system is 
+loaded beyond 50%.
+Systems that don't use SMT will degrade as well, to some extent, because of shared resource contention.
+In the following the extent of this degradation is investigated with a few illustrative workloads.
+
+
+### Experimental Setup
+
+The test systems are limited to eight cores to limit the maximum packet load needed to saturate them. 
+This is to exclude the influence of the networking infrastructure on the experiment.
+All systems run Amazon Linux 2 with Linux kernel 5.10.118.
+The test system are virtualized instances that use S3 networked storage.
+
+
+#### First experiment
+
+The systems to be tested are an 8 vCPU Intel instance (8375C, 3.5GHz), c6i.2xlarge, and an 8 CPU Graviton 3 instance, c7g.2xlarge.
+OpenSSL source is placed and configured in eight separate directories.
+The test then compiles OpenSSL one to eight times in parallel, limited to one CPU per compile.
+An ideal system would take the exact same time for each compile run.
+
+Results:
+
+```
+Graviton 3 System
+ 
+Number of compiles, CPU use [%], time to complete [s]
+
+1, 16, 244
+2, 26, 249
+3, 38, 247
+4, 51, 250
+5, 63, 251
+6, 75, 255
+7, 88, 251
+8, 100, 253
+```
+
+The Graviton 3 achieves 96% of the single compile time performance, single CPU baseline when using 100% of the available CPUs.
+244/253=0.964
+
+
+```
+Intel SMT system
+
+Number of compiles, CPU use [%], time to complete [s]
+
+1, 20, 223
+2, 26, 226
+3, 38, 224
+4, 51, 225
+5, 63, 265
+6, 76, 301
+7, 88, 327
+8, 100, 354
+```
+
+
+The Intel system achieves 63% of the compile time performance of the single compile, single CPU baseline when loaded to 100%.
+223/354=0.63
+![](images/system-load/GccC7gC6i.png)
+
+Note: 
+The average compile time of the c6i instance is 268s, close to the 250s on c7g. 
+The Linux scheduler picks cores over threads, when available, which shows nicely in the steep increase in compile time of the c6i beyond 50% load.
+
+#### Second experiment
```

Comment:
This is the most interesting experiment for our customers.

Reply:
I put this information in the introduction, as well.